### DISCUSSION TITLE
Issue with trigger after DST

### DISCUSSION THREAD

**Discussion by arunprakashn** [quartznet/quartznet#1441] [Q&A]
Note: I have already gone through the FAQ section on DST impact with respect to the simple and cron jobs. I still have a problem.

We are using a simple schedule with 5-second intervals. If I move my system clock to 1 hour before manually to simulate the DST, The job trigger stops. It resumes only after an hour. I am sure that we are not using Cron schedule. Can anyone tell why this is happening?
```
 triggerBuilder.WithSimpleSchedule(s => s.WithIntervalInSeconds(intervalInSeconds).RepeatForever().WithMisfireHandlingInstructionNowWithExistingCount());

```
v3.2.3

To reproduce: Set up a simple schedule for 5 seconds. Print a console message when the job is fired. Change the time manually one hour back. You will notice that the triggers will stop.

**Comment by lahma** [comment]
Quartz calculates next fire time for trigger and waits for it. If you manually move the clock backwards it will take an hour to reach the pre-calculated fire time. You can experiment with calculation logic by using `TriggerUtils.ComputeFireTimesBetween`, but you should use a time span that has actual DST change in it.

**Reply by arunprakashn** [reply]
Hi @lahma ,

Please let me know if my understanding is correct. We have an application set up for a simple scheduler. It is to be triggered every 5 sec. If the DST happens and the time goes 1 hour backwards, are you saying, it wont impact the trigger? We faced some issue and I was trying to replicate the DST issue by moving my clock manually backwards and observed that the trigger stopped firing. You mean to say that the DST change of 1 hour backward and manual time change of 1 hour backwards are treated differently by the Quartz? In essence, you are confirming that the DST timeshift wont impact the simple 5 sec schedule. Is that correct?

**Reply by lahma** [reply]
Please use `TriggerUtils.ComputeFireTimesBetween` and print out the fire times to check how you trigger will behave in the given time range.

**Comment by GieltjE** [comment]
There still appears to be an issue every few time changes (and isn't just DST related), even when you expect the triggers to fire.
We use the same builder as mentioned above and rectify the system time every few hours.

There is a reliable way to trigger this:
1: create a trigger that fires every few seconds like the one above
2: start the program
3: set the system date to yesterday (don't change the time)
4: wait just a few seconds
5: set the system date back to today (don't change the time)
6: observe the job no longer getting triggered even though we are back in a logical time

As a slight complication, we cannot trust the "SystemEvents.TimeChanged" event, this one will simply not fire when running in a service (and in a few other undocumented scenarios, however running a form and overwriting WndProc and catching WM_TIMECHANGE works).
What is the best way to reset everything when the time changed event happens?

**Reply by GieltjE** [reply]
@lahma can we get some feedback on this issue?
I'm fully aware that this isn't going to be an easy fix, but triggers dying when a cheap quartz crystal on the motherboard gets rectified (NTP or manual) doesn't really sound right.

**Reply by lahma** [reply]
Maybe a failing test case would demonstrate the problem better and also offer easier startup point for investigation. Quartz has `SystemTime` which allows you to change time inside the test case.

**Reply by GieltjE** [reply]
There is now a test case #1490, only works for Windows as of now (don't know of any cross platform system time change code at the moment).

**Comment by lahma** [comment]
Closing the loop on this one — there was a real bug behind it, now fixed on both 3.x and 4.x.

Two separate things were going on:

**1. Expected behaviour.** Quartz schedules against the wall clock. Moving the clock back an hour means a trigger whose next fire time was already computed won't fire until the clock has caught up — that hour has to be lived through twice. That part is by design and applies to real DST transitions too.

**2. An actual bug.** Several sleep durations were computed by subtracting two *wall-clock* timestamps and then passed to a *monotonic* wait (`Monitor.Wait` / `Task.Delay`). Move the clock back by a day and the computed duration inflates by a day, so the thread sleeps for a day of **real** time — it never woke up even after the clock was restored. @EbenZhang nailed this in https://github.com/quartznet/quartznet/issues/1508#issuecomment-1057882900, and @GieltjE / @DirkieDurky's reproduction in #1490 is what made it reproducible.

Three places had it:

- `QuartzSchedulerThread`'s trigger wait — triggers stopped firing. Fixed in #2929.
- `MisfireHandler` — misfire recovery stopped. Fixed now (3.x: #3147, 4.x: #3148).
- `ClusterManager` — check-in stopped, so peer nodes would declare the instance failed and recover its triggers. Fixed now (3.x: #3147, 4.x: #3148).

**What to expect now:** recovery is bounded but not instantaneous. After a clock correction the firing loop resumes within one `quartz.scheduler.idleWaitTime` (30 s default); with AdoJobStore, misfire handling resumes within one `misfireHandlerFrequency` and cluster check-in within one `clusterCheckinInterval`.

If you want to test clock changes yourself, fake the clock rather than changing the machine clock (`SystemTime.UtcNow` on 3.x, a custom `TimeProvider` on 4.x) — that isolates the wall clock from monotonic timers, which is exactly the condition that surfaces this class of bug.
